include/deal.II/base/vectorization.h

```diff
@@ -255,31 +256,29 @@ class VectorizedArrayBase
   /**
    * Default constructor.
    */
-  VectorizedArrayBase() = default;
+  constexpr VectorizedArrayBase() = default;
 
   /**
    * Construct an array with the given initializer list.
+   *
+   * The initializer list must have at most as many elements as the
+   * vector length. Elements not listed in the initializer list are
+   * zero-initialized.
    */
   template <typename U>
-  VectorizedArrayBase(const std::initializer_list<U> &list)
+  constexpr VectorizedArrayBase(const std::initializer_list<U> &list)
   {
-    auto i0 = this->begin();
-    auto i1 = list.begin();
-
-    for (; i1 != list.end(); ++i0, ++i1)
-      {
-        Assert(
-          i0 != this->end(),
-          ExcMessage(
-            "Initializer list exceeds size of this VectorizedArray object."));
+    const unsigned int n_initializers = list.size();
+    Assert(n_initializers,
+           ExcMessage("The initializer list must have at most "
+                      "as many elements as the vector length."));
 
-        *i0 = *i1;
-      }
+    // Copy what's in the list.
+    std::copy(list.begin(), list.end(), this->begin());
 
-    for (; i0 != this->end(); ++i0)
-      {
-        *i0 = 0.0;
-      }
+    // Then add zero padding where necessary.
+    if (list.size() < size())
+      std::fill(this->begin() + n_initializers, this->end(), 0.0);
   }
```

Comment:
I chose to clean this function up a bit. I think it is easier to read if we move the assertion out of the loop.